{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# `Project Outline Berlin`\n",
    "\n",
    "\n",
    "## Objective:\n",
    "\n",
    "> Make a `Snakemake pipeline` that outputs a dataframe or an Excel sheet that show which amino acid is present at a given position per lineage and segment of influenza viruses. Hereby trying to make a somewhat more flexible tool, so it could also be applied to other lineages than used here such as H5N1 or H7N9 lineages.\n",
    ">\n",
    "> FASTA files may contain nucloetide seqeunces from segments of influenza virus subtypes `H1N1pdm`, `H3N2`, `B-Vic` and `B-Yam` downloaded from GISAID.\n",
    ">\n",
    "> This FASTA file will first be processed by `Nextclade` (for `HA` & `NA` segments) and `Nextalign` (for `PB2`, `PB1`, `PA`, `NP`, `MA` & `NS` segments). Here, input nucleotide sequences are aligned to their appropriate reference genomes. Translated alignments are truncated to the sequence length representing the viral functional protein.\n",
    "> \n",
    "> `Nextclade` and `Nextalign` both output aligned nucleotide sequences and aligned amino acid sequences that are truncated according to a correct numbering used for annotation.\n",
    ">\n",
    ">   **Reference for HA numbering:**<br/>\n",
    ">Burke DF and Smith DJ (2014) <br/>A standardised numbering for all subtypes of Influenza A hemaggluttin (HA) sequences. <br/>Plos One 9(11): e112302 <br/>DOI: 10.1371/journal.pone.0112302 <br/>[Reference table](https://antigenic-cartography.org/surveillance/evergreen/HAnumbering/)\n",
    ">\n",
    "> Python scripts will than process the output files from `Nextclade` and `Nextalign` and finally output tables where rows are represented by the input sequence names and where columns represent correctly numbered amino acids posititions within the analysed segment."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Input and output files\n",
    "\n",
    "#### INPUT FILES\n",
    "\n",
    "1. Downloaded FASTA files from GISAID with nucleotide sequences and headers that are formatted like:\n",
    "\n",
    "```\n",
    "Isolate ID | Isolate name | Type | Lineage | Segment | Passage details/history | Collection date\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "**Example fasta header:**    \n",
    "```\n",
    ">EPI_ISL_342146|A/Netherlands/00322/2019|A_/_H1N1|pdm09|HA|Original|2019-02-06\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "2. Lists of amino acid positions per viral protein sequences grouped by subtype and gene segment.       \n",
    "    Applying an Excel spreadsheet as a provided input might give more flexibility in use of the final pipeline.\n",
    "\n",
    "\n",
    "    *Example input table:*\n",
    "\n",
    "lineage | segment | position\n",
    "--------|---------|---------\n",
    "H1N1pdm | HA      | 156\n",
    "H1N1pdm | HA      | 158\n",
    "H1N1pdm | HA      | 162\n",
    "H1N1pdm | NA      | 274\n",
    "H1N1pdm | PA      | 31\n",
    "H3N2    | HA      | 145\n",
    "H3N2    | HA      | 184\n",
    "H3N2    | HA      | 185\n",
    "etc...  | etc...  | etc...\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "3. Datasets that are used for `Nextclade` and `Nextalign`to run:\n",
    "\n",
    "- **Nextalign input files:**\n",
    "    - `_nextalign/data/flu_h1n1pdm_pa/genemap.gff`\n",
    "    - `_nextalign/data/flu_h1n1pdm_pa/reference.fasta`\n",
    "    - `_nextalign/data/flu_h1n1pdm_pa/sequences.fasta`\n",
    "\n",
    "<br/>\n",
    "\n",
    "- **Nextclade input files:**\n",
    "    - `_nextclade/data/flu_h1n1pdm_ha/primers.cvs`\n",
    "    - `_nextclade/data/flu_h1n1pdm_ha/genemap.gff`\n",
    "    - `_nextclade/data/flu_h1n1pdm_ha/qc.json`\n",
    "    - `_nextclade/data/flu_h1n1pdm_ha/reference.fasta`\n",
    "    - `_nextclade/data/flu_h1n1pdm_ha/sequences.fasta`\n",
    "    - `_nextclade/data/flu_h1n1pdm_ha/tag.json`\n",
    "    - `_nextclade/data/flu_h1n1pdm_ha/tree.json`\n",
    "    - `_nextclade/data/flu_h1n1pdm_ha/virus_properties.json`\n",
    "\n",
    "\n",
    "\n",
    "#### OUTPUT FILES\n",
    "\n",
    "- `Nextclade` and `Nextalign` should give outputs in defined folders that are named in convention with their lineage and segment (`output/Nextclade/H1N1/PB1/`).\n",
    "\n",
    "- **Nextclade output files:**\n",
    "    - `output/Nextclade/H1N1/HA/nextclade.csv` (Contains additional clade information)\n",
    "    - `output/Nextclade/H1N1/HA/nextclade.aligned.fasta.fasta`\n",
    "    - `output/Nextclade/H1N1/HA/nextclade_gene_HA1.translation.fasta`\n",
    "    - `output/Nextclade/H1N1/HA/nextclade_gene_HA2.translation.fasta`\n",
    "    - `output/Nextclade/H1N1/HA/nextclade_gene_SigPep.translation.fasta.fasta`\n",
    "\n",
    "<br/>\n",
    "\n",
    " - **Nextalign output files:**\n",
    "    - `output/Nextalign/H1N1/PA/nextalign.aligned.fasta`\n",
    "    - `output/Nextalign/H1N1/PA/nextalign_gene_PA.translation.fasta`\n",
    "\n",
    "\n",
    "\n",
    "2. Dataframe or Excel sheet with listing what amino acids are present for given positions per sequence in the final protein alignments. For HA and NA clades could be included to complement the output data of AA's per position."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## NOTE | Conda Configuration\n",
    "\n",
    "### Conda set-up within arm-64 installation of conda.\n",
    "Conda packages are installed within an osx-arm64 architecture (Intel) subdirectory of conda installation that runs osx-arm64 architecture (Apple M1/2).\n",
    "In order for the `Snakemake`, `Nextclade` and `Nextalign` packages to work, they should run using a Conda environment that supports osx-64 packages.\n",
    "\n",
    "This is where we tell Conda that we want our environment to use x86 architecture rather than the native ARM64 architecture. We can do that with following lines of code, which create and activate a new conda environment called `berlin`:\n",
    "\n",
    "```\n",
    "CONDA_SUBDIR=osx-64 conda create --name berlin python=3.10\n",
    "conda activate berlin\n",
    "conda config --env --set subdir osx-64\n",
    "```\n",
    "\n",
    "- The first line creates the environment. We simply set the CONDA_SUBDIR environment variable to indicate that conda should create the ennvironment with an osx-64 Python executable. \n",
    "- The second line activates the environment, and the third line ensures that conda installs osx-64 versions of packages into the environment.\n",
    "\n",
    "> This new environment will now be able to use dependencies with osx-64 builds."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conda packages\n",
    "\n",
    "#### Git:\n",
    "```\n",
    "conda install -c anaconda git\n",
    "```\n",
    "\n",
    "#### Jupyter:\n",
    "```\n",
    "conda install -c anaconda jupyter\n",
    "```\n",
    "\n",
    "#### Snakemake (osx-64):\n",
    "```\n",
    "conda install -c bioconda snakemake\n",
    "```\n",
    "\n",
    "#### Nextalign (osx-64):\n",
    "```\n",
    "conda install -c bioconda nextalign\n",
    "```\n",
    "\n",
    "#### Nextclade (osx-64):\n",
    "```\n",
    "conda install -c bioconda nextclade  \n",
    "```\n",
    "\n",
    "#### Graphviz (extension for Snakemake):\n",
    "Helper tool to visualize the DAG (Directed Acyclic Graph) and save it as a .png file.\n",
    "\n",
    "*Source:*\n",
    "- Youtube | [An introduction to Snakemake tutorial for beginners (CC248)](https://www.youtube.com/watch?v=r9PWnEmz_tc)\n",
    "\n",
    "```\n",
    "conda install graphviz\n",
    "```\n",
    "\n",
    "To use the tool on the Snakemake file, run:\n",
    "\n",
    "```\n",
    "snakemake --dag targets | dot Tpng > dag.png\n",
    "```\n",
    "\n",
    "> NOTE: In the video the ```rule targets``` was made, listing all the outputs that need to be generated when running the basic snakemake command: \n",
    "\n",
    "```snakemake -c 1``` (runs using one thread)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Code chunks\n",
    "\n",
    "## Retrieve positions of interest for given lineages and seqments.\n",
    "- Use pandas to read an excel file that lists positions per lineage and segment.\n",
    "\n",
    "    Reminder: the file `input/positions_by_lineage_and_segment.xlsx` has the folowing columns:\n",
    "\n",
    "lineage | segment | position\n",
    "--------|---------|---------\n",
    "H1N1pdm | HA      | 156\n",
    "H1N1pdm | HA      | 158\n",
    "H1N1pdm | HA      | 162\n",
    "H1N1pdm | NA      | 274\n",
    "H1N1pdm | PA      | 31\n",
    "H3N2    | HA      | 145\n",
    "H3N2    | HA      | 184\n",
    "H3N2    | HA      | 185\n",
    "etc...  | etc...  | etc...\n",
    "\n",
    "<br>\n",
    "\n",
    "- Import data from `input/positions_by_lineage_and_segments.xlsx` to get the desired positions to look for, within the AA sequence for a given lineage and segment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libraries:\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "#########################################################################################\n",
    "#                                                                                       #\n",
    "# FUNCTION: Get headers and amino acid sequences from Nextaligns output files.          #\n",
    "#                                                                                       #\n",
    "#########################################################################################\n",
    "\n",
    "def get_sequence_names_and_amino_acid_sequences(fasta_file):\n",
    "\n",
    "    # Read the FASTA file and extract the nucleotide sequences.\n",
    "    records = list(SeqIO.parse(fasta_file, \"fasta\"))\n",
    "\n",
    "    # If there are no sequences in the fasta file, abort the function.\n",
    "    if len(records) < 1:\n",
    "        print(\"No sequences found in the FASTA file.\")\n",
    "        return\n",
    "\n",
    "    # Get a list of sequence names from the headers.\n",
    "    sequence_names = []\n",
    "    for record in records:\n",
    "        sequence_names.append(record.id)\n",
    "\n",
    "    # Get a list of corresponding amino acid sequences\n",
    "    amino_acid_sequences = []\n",
    "    for record in records:\n",
    "        amino_acid_sequences.append(record.seq)\n",
    "\n",
    "    return sequence_names, amino_acid_sequences\n",
    "\n",
    "\n",
    "#########################################################################################\n",
    "#                                                                                       #\n",
    "# FUNCTION: Get a list of positions of interest for lineage and segment of interest.    #\n",
    "#                                                                                       #\n",
    "#########################################################################################\n",
    "\n",
    "def get_positions_from_excel(excel_file, lineage, segment):\n",
    "\n",
    "    # Import pandas\n",
    "    import pandas as pd\n",
    "\n",
    "    # Use pandas to read columns from the Excel file\n",
    "    position_by_lineage_and_segment = pd.read_excel(excel_file)\n",
    "\n",
    "    # Make a 'selection' by filtering data to match positions for 'ha' segments belonging to the 'vic'lineage.\n",
    "    selection = position_by_lineage_and_segment.query(\n",
    "        f'lineage==\"{lineage}\" & segment==\"{segment}\"')\n",
    "\n",
    "    # Get the selected values form the 'positions' columns as a list.\n",
    "    positions = selection['position'].to_list()\n",
    "\n",
    "    return positions\n",
    "\n",
    "\n",
    "#########################################################################################\n",
    "#                                                                                       #\n",
    "# FUNCTION: Make an Excel file listing headers with amino acids at given positions.     #\n",
    "#                                                                                       #\n",
    "#########################################################################################\n",
    "\n",
    "def generate_excel_output(output, sequence_names, amino_acid_sequences, positions):\n",
    "\n",
    "    data = []\n",
    "    for sequence_name, amino_acid_sequence in zip(sequence_names, amino_acid_sequences):\n",
    "        row = [sequence_name]\n",
    "        row.extend(amino_acid_sequence[pos - 1] for pos in positions)\n",
    "        data.append(row)\n",
    "\n",
    "    column_names = [\"Sequence\"] + [f\"Pos {pos}\" for pos in positions]\n",
    "\n",
    "    df = pd.DataFrame(data, columns=column_names)\n",
    "\n",
    "    df.to_excel(output, index=False)\n",
    "\n",
    "\n",
    "#########################################################################################\n",
    "#                                                                                       #\n",
    "#  Run the functions above...                                                           #\n",
    "#                                                                                       #\n",
    "#########################################################################################\n",
    "\n",
    "\n",
    "\n",
    "#\n",
    "# -----------------------------    HEADERS & AA SEQUENCES   ----------------------------\n",
    "#\n",
    "# Set input fasta file (SNAKEFILE):\n",
    "aa_fasta = '_nextclade/output/flu_vic_ha/nextclade_gene_HA1.translation.fasta'\n",
    "\n",
    "# Get 'aa_names' and 'aa_sequences' from input fasta file:\n",
    "aa_names, aa_sequences = get_sequence_names_and_amino_acid_sequences(\n",
    "                            fasta_file=aa_fasta\n",
    "                            )\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#\n",
    "# --------------------------    FILTER POSITIONS OF INTEREST   ------------------------\n",
    "#\n",
    "# Set 'positions' using the input Excel file and by defining filters (SNAKEFILE):\n",
    "positions_table = 'input/positions_by_lineage_and_segment.xlsx'\n",
    "\n",
    "filter_lineage = 'vic'\n",
    "\n",
    "filter_segment = 'ha'\n",
    "\n",
    "# Get the input positions from the input excel_file that are filtered by lineage and segment:\n",
    "filtered_positions = get_positions_from_excel(\n",
    "                        excel_file=positions_table,\n",
    "                        lineage=filter_lineage,\n",
    "                        segment=filter_segment\n",
    "                        )\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#\n",
    "# -------------------------    CREATE OUTPUT EXCEL FILE   -----------------------------\n",
    "#\n",
    "# Set the name of the final output file (SNAKEFILE):\n",
    "excel_output = 'output/aa_at_positions_for_vic_HA1.xlsx'\n",
    "\n",
    "# Create an Excel file as final output:\n",
    "if aa_names and aa_sequences:\n",
    "    generate_excel_output(\n",
    "        output=excel_output,\n",
    "        sequence_names=aa_names,\n",
    "        amino_acid_sequences=aa_sequences,\n",
    "        positions=filtered_positions\n",
    "    )\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "vcs",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
