{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# `Project Outline Berlin`\n",
    "\n",
    "\n",
    "## Objective:\n",
    "\n",
    "> Make a `Snakemake pipeline` that outputs a dataframe or an Excel sheet that show which amino acid is present at a given position per lineage and segment of influenza viruses. Hereby trying to make a somewhat more flexible tool, so it could also be applied to other lineages than used here such as H5N1 or H7N9 lineages.\n",
    ">\n",
    "> FASTA files may contain nucloetide seqeunces from segments of influenza virus subtypes `H1N1pdm`, `H3N2`, `B-Vic` and `B-Yam` downloaded from GISAID.\n",
    ">\n",
    "> This FASTA file will first be processed by `Nextclade` (for `HA` & `NA` segments) and `Nextalign` (for `PB2`, `PB1`, `PA`, `NP`, `MA` & `NS` segments). Here, input nucleotide sequences are aligned to their appropriate reference genomes. Translated alignments are truncated to the sequence length representing the viral functional protein.\n",
    "> \n",
    "> `Nextclade` and `Nextalign` both output aligned nucleotide sequences and aligned amino acid sequences that are truncated according to a correct numbering used for annotation.\n",
    ">\n",
    ">   **Reference for HA numbering:**<br/>\n",
    ">Burke DF and Smith DJ (2014) <br/>A standardised numbering for all subtypes of Influenza A hemaggluttin (HA) sequences. <br/>Plos One 9(11): e112302 <br/>DOI: 10.1371/journal.pone.0112302 <br/>[Reference table](https://antigenic-cartography.org/surveillance/evergreen/HAnumbering/)\n",
    ">\n",
    "> Python scripts will than process the output files from `Nextclade` and `Nextalign` and finally output tables where rows are represented by the input sequence names and where columns represent correctly numbered amino acids posititions within the analysed segment."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Input and output files\n",
    "\n",
    "#### INPUT FILES\n",
    "\n",
    "1. Downloaded FASTA files from GISAID with nucleotide sequences and headers that are formatted like:\n",
    "\n",
    "```\n",
    "Isolate ID | Isolate name | Type | Lineage | Segment | Passage details/history | Collection date\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "**Example fasta header:**    \n",
    "```\n",
    ">EPI_ISL_342146|A/Netherlands/00322/2019|A_/_H1N1|pdm09|HA|Original|2019-02-06\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "2. Lists of amino acid positions per viral protein sequences grouped by subtype and gene segment.       \n",
    "    Applying an Excel spreadsheet as a provided input might give more flexibility in use of the final pipeline.\n",
    "\n",
    "\n",
    "    *Example input table:*\n",
    "\n",
    "lineage | segment | position\n",
    "--------|---------|---------\n",
    "H1N1pdm | HA      | 156\n",
    "H1N1pdm | HA      | 158\n",
    "H1N1pdm | HA      | 162\n",
    "H1N1pdm | NA      | 274\n",
    "H1N1pdm | PA      | 31\n",
    "H3N2    | HA      | 145\n",
    "H3N2    | HA      | 184\n",
    "H3N2    | HA      | 185\n",
    "etc...  | etc...  | etc...\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "3. Datasets that are used for `Nextclade` and `Nextalign`to run:\n",
    "\n",
    "- **Nextalign input files:**\n",
    "    - `_nextalign/data/flu_h1n1pdm_pa/genemap.gff`\n",
    "    - `_nextalign/data/flu_h1n1pdm_pa/reference.fasta`\n",
    "    - `_nextalign/data/flu_h1n1pdm_pa/sequences.fasta`\n",
    "\n",
    "<br/>\n",
    "\n",
    "- **Nextclade input files:**\n",
    "    - `_nextclade/data/flu_h1n1pdm_ha/primers.cvs`\n",
    "    - `_nextclade/data/flu_h1n1pdm_ha/genemap.gff`\n",
    "    - `_nextclade/data/flu_h1n1pdm_ha/qc.json`\n",
    "    - `_nextclade/data/flu_h1n1pdm_ha/reference.fasta`\n",
    "    - `_nextclade/data/flu_h1n1pdm_ha/sequences.fasta`\n",
    "    - `_nextclade/data/flu_h1n1pdm_ha/tag.json`\n",
    "    - `_nextclade/data/flu_h1n1pdm_ha/tree.json`\n",
    "    - `_nextclade/data/flu_h1n1pdm_ha/virus_properties.json`\n",
    "\n",
    "\n",
    "\n",
    "#### OUTPUT FILES\n",
    "\n",
    "- `Nextclade` and `Nextalign` should give outputs in defined folders that are named in convention with their lineage and segment (`output/Nextclade/H1N1/PB1/`).\n",
    "\n",
    "- **Nextclade output files:**\n",
    "    - `output/Nextclade/H1N1/HA/nextclade.csv` (Contains additional clade information)\n",
    "    - `output/Nextclade/H1N1/HA/nextclade.aligned.fasta.fasta`\n",
    "    - `output/Nextclade/H1N1/HA/nextclade_gene_HA1.translation.fasta`\n",
    "    - `output/Nextclade/H1N1/HA/nextclade_gene_HA2.translation.fasta`\n",
    "    - `output/Nextclade/H1N1/HA/nextclade_gene_SigPep.translation.fasta.fasta`\n",
    "\n",
    "<br/>\n",
    "\n",
    " - **Nextalign output files:**\n",
    "    - `output/Nextalign/H1N1/PA/nextalign.aligned.fasta`\n",
    "    - `output/Nextalign/H1N1/PA/nextalign_gene_PA.translation.fasta`\n",
    "\n",
    "\n",
    "\n",
    "2. Dataframe or Excel sheet with listing what amino acids are present for given positions per sequence in the final protein alignments. For HA and NA clades could be included to complement the output data of AA's per position."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## NOTE  |  Conda subdirectory for `osx-64` packages\n",
    "\n",
    "### Conda set-up within osx-arm64 installation of conda.\n",
    "Conda packages are installed within an osx-arm64 architecture (Intel) subdirectory of conda installation that runs osx-arm64 architecture (Apple M1/2).\n",
    "In order for the `Snakemake`, `Nextclade` and `Nextalign` packages to work, they should run using a Conda environment that supports osx-64 packages.\n",
    "\n",
    "This is where we tell Conda that we want our environment to use x86 architecture rather than the native ARM64 architecture. We can do that with following lines of code, which create and activate a new conda environment called `berlin`:\n",
    "\n",
    "```\n",
    "CONDA_SUBDIR=osx-64 conda create --name berlin python=3.10\n",
    "conda activate berlin\n",
    "conda config --env --set subdir osx-64\n",
    "```\n",
    "\n",
    "- The first line creates the environment. We simply set the CONDA_SUBDIR environment variable to indicate that conda should create the ennvironment with an osx-64 Python executable. \n",
    "- The second line activates the environment, and the third line ensures that conda installs osx-64 versions of packages into the environment.\n",
    "\n",
    "> This new environment will now be able to use dependencies with osx-64 builds."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conda packages\n",
    "\n",
    "#### Git:\n",
    "```\n",
    "conda install -c anaconda git\n",
    "```\n",
    "\n",
    "#### Jupyter:\n",
    "```\n",
    "conda install -c anaconda jupyter\n",
    "```\n",
    "\n",
    "#### Snakemake (osx-64):\n",
    "```\n",
    "conda install -c bioconda snakemake\n",
    "```\n",
    "\n",
    "#### Nextalign (osx-64):\n",
    "```\n",
    "conda install -c bioconda nextalign\n",
    "```\n",
    "\n",
    "#### Nextclade (osx-64):\n",
    "```\n",
    "conda install -c bioconda nextclade  \n",
    "```\n",
    "\n",
    "#### Graphviz (extension for Snakemake):\n",
    "Helper tool to visualize the DAG (Directed Acyclic Graph) and save it as a .png file.\n",
    "\n",
    "*Source:*\n",
    "- Youtube | [An introduction to Snakemake tutorial for beginners (CC248)](https://www.youtube.com/watch?v=r9PWnEmz_tc)\n",
    "\n",
    "```\n",
    "conda install graphviz\n",
    "```\n",
    "\n",
    "To use the tool on the Snakemake file, run:\n",
    "\n",
    "```\n",
    "snakemake --dag targets | dot Tpng > dag.png\n",
    "```\n",
    "\n",
    "#### *A better working option:*\n",
    "```\n",
    "snakemake --forceall --rulegraph | dot -Tpdf > dag.pdf\n",
    "```\n",
    "\n",
    "\n",
    "> NOTE: In the video the ```rule targets``` was made, listing all the outputs that need to be generated when running the basic snakemake command: \n",
    "\n",
    "```snakemake -c 1``` (runs using one thread)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Code chunks\n",
    "\n",
    "## Retrieve positions of interest for given lineages and seqments.\n",
    "- Use pandas to read an excel file that lists positions per lineage and segment.\n",
    "\n",
    "    Reminder: the file `input/positions_by_lineage_and_segment.xlsx` has the folowing columns:\n",
    "\n",
    "lineage | segment | position\n",
    "--------|---------|---------\n",
    "H1N1pdm | HA      | 156\n",
    "H1N1pdm | HA      | 158\n",
    "H1N1pdm | HA      | 162\n",
    "H1N1pdm | NA      | 274\n",
    "H1N1pdm | PA      | 31\n",
    "H3N2    | HA      | 145\n",
    "H3N2    | HA      | 184\n",
    "H3N2    | HA      | 185\n",
    "etc...  | etc...  | etc...\n",
    "\n",
    "<br>\n",
    "\n",
    "- Import data from `input/positions_by_lineage_and_segments.xlsx` to get the desired positions to look for, within the AA sequence for a given lineage and segment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libraries:\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "#########################################################################################\n",
    "#                                                                                       #\n",
    "# FUNCTION: Get headers and amino acid sequences from Nextaligns output files.          #\n",
    "#                                                                                       #\n",
    "#########################################################################################\n",
    "\n",
    "def get_sequence_names_and_amino_acid_sequences(fasta_file):\n",
    "\n",
    "    # Read the FASTA file and extract the nucleotide sequences.\n",
    "    records = list(SeqIO.parse(fasta_file, \"fasta\"))\n",
    "\n",
    "    # If there are no sequences in the fasta file, abort the function.\n",
    "    if len(records) < 1:\n",
    "        print(\"No sequences found in the FASTA file.\")\n",
    "        return\n",
    "\n",
    "    # Get a list of sequence names from the headers.\n",
    "    sequence_names = []\n",
    "    for record in records:\n",
    "        sequence_names.append(record.id)\n",
    "\n",
    "    # Get a list of corresponding amino acid sequences\n",
    "    amino_acid_sequences = []\n",
    "    for record in records:\n",
    "        amino_acid_sequences.append(record.seq)\n",
    "\n",
    "    return sequence_names, amino_acid_sequences\n",
    "\n",
    "\n",
    "#########################################################################################\n",
    "#                                                                                       #\n",
    "# FUNCTION: Get a list of positions of interest for lineage and segment of interest.    #\n",
    "#                                                                                       #\n",
    "#########################################################################################\n",
    "\n",
    "def get_positions_from_excel(excel_file, lineage, segment):\n",
    "\n",
    "    # Import pandas\n",
    "    import pandas as pd\n",
    "\n",
    "    # Use pandas to read columns from the Excel file\n",
    "    position_by_lineage_and_segment = pd.read_excel(excel_file)\n",
    "\n",
    "    # Make a 'selection' by filtering data to match positions for 'ha' segments belonging to the 'vic'lineage.\n",
    "    selection = position_by_lineage_and_segment.query(\n",
    "        f'lineage==\"{lineage}\" & segment==\"{segment}\"')\n",
    "\n",
    "    # Get the selected values form the 'positions' columns as a list.\n",
    "    positions = selection['position'].to_list()\n",
    "\n",
    "    return positions\n",
    "\n",
    "\n",
    "#########################################################################################\n",
    "#                                                                                       #\n",
    "# FUNCTION: Make an Excel file listing headers with amino acids at given positions.     #\n",
    "#                                                                                       #\n",
    "#########################################################################################\n",
    "\n",
    "def generate_excel_output(output, sequence_names, amino_acid_sequences, positions):\n",
    "\n",
    "    data = []\n",
    "    for sequence_name, amino_acid_sequence in zip(sequence_names, amino_acid_sequences):\n",
    "        row = [sequence_name]\n",
    "        row.extend(amino_acid_sequence[pos - 1] for pos in positions)\n",
    "        data.append(row)\n",
    "\n",
    "    column_names = [\"Sequence\"] + [f\"Pos {pos}\" for pos in positions]\n",
    "\n",
    "    df = pd.DataFrame(data, columns=column_names)\n",
    "\n",
    "    df.to_excel(output, index=False)\n",
    "\n",
    "\n",
    "#########################################################################################\n",
    "#                                                                                       #\n",
    "#  Run the functions above...                                                           #\n",
    "#                                                                                       #\n",
    "#########################################################################################\n",
    "\n",
    "\n",
    "\n",
    "#\n",
    "# -----------------------------    HEADERS & AA SEQUENCES   ----------------------------\n",
    "#\n",
    "# Set input fasta file (SNAKEFILE):\n",
    "aa_fasta = '_nextclade/output/flu_vic_ha/nextclade_gene_HA1.translation.fasta'\n",
    "\n",
    "# Get 'aa_names' and 'aa_sequences' from input fasta file:\n",
    "aa_names, aa_sequences = get_sequence_names_and_amino_acid_sequences(\n",
    "                            fasta_file=aa_fasta\n",
    "                            )\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#\n",
    "# --------------------------    FILTER POSITIONS OF INTEREST   ------------------------\n",
    "#\n",
    "# Set 'positions' using the input Excel file and by defining filters (SNAKEFILE):\n",
    "positions_table = 'input/positions_by_lineage_and_segment.xlsx'\n",
    "\n",
    "filter_lineage = 'vic'\n",
    "\n",
    "filter_segment = 'ha'\n",
    "\n",
    "# Get the input positions from the input excel_file that are filtered by lineage and segment:\n",
    "filtered_positions = get_positions_from_excel(\n",
    "                        excel_file=positions_table,\n",
    "                        lineage=filter_lineage,\n",
    "                        segment=filter_segment\n",
    "                        )\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#\n",
    "# -------------------------    CREATE OUTPUT EXCEL FILE   -----------------------------\n",
    "#\n",
    "# Set the name of the final output file (SNAKEFILE):\n",
    "excel_output = 'output/aa_at_positions_for_vic_HA1.xlsx'\n",
    "\n",
    "# Create an Excel file as final output:\n",
    "if aa_names and aa_sequences:\n",
    "    generate_excel_output(\n",
    "        output=excel_output,\n",
    "        sequence_names=aa_names,\n",
    "        amino_acid_sequences=aa_sequences,\n",
    "        positions=filtered_positions\n",
    "    )\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Looking at the output of the function `input_func_all()` using a dictionary matching segments and genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'genes_translated': ['output/vic/pa/nextalign_gene_PA.translation.fasta', 'output/h1n1pdm/np/nextalign_gene_NP.translation.fasta', 'output/vic/ha/nextalign_gene_HA1.translation.fasta', 'output/h3n2/na/nextalign_gene_NA.translation.fasta', 'output/h1n1pdm/ns/nextalign_gene_NS2.translation.fasta', 'output/vic/na/nextalign_gene_NA.translation.fasta', 'output/h3n2/pa/nextalign_gene_PA.translation.fasta', 'output/h3n2/ha/nextalign_gene_HA1.translation.fasta', 'output/h1n1pdm/ma/nextalign_gene_MA.translation.fasta', 'output/h1n1pdm/ha/nextalign_gene_HA1.translation.fasta', 'output/h1n1pdm/pa/nextalign_gene_PA.translation.fasta', 'output/vic/np/nextalign_gene_NP.translation.fasta', 'output/h1n1pdm/pb2/nextalign_gene_PB2.translation.fasta', 'output/h3n2/ma/nextalign_gene_MA.translation.fasta', 'output/h3n2/pb1/nextalign_gene_PB1.translation.fasta', 'output/vic/pb2/nextalign_gene_PB2.translation.fasta', 'output/h3n2/ns/nextalign_gene_NS2.translation.fasta', 'output/vic/ns/nextalign_gene_NS2.translation.fasta', 'output/h1n1pdm/pb1/nextalign_gene_PB1.translation.fasta', 'output/h1n1pdm/na/nextalign_gene_NA.translation.fasta', 'output/h3n2/np/nextalign_gene_NP.translation.fasta', 'output/vic/pb1/nextalign_gene_PB1.translation.fasta', 'output/h3n2/pb2/nextalign_gene_PB2.translation.fasta', 'output/vic/ma/nextalign_gene_MA.translation.fasta'], 'nt_segments_aligned': ['output/vic/pa/nextalign.aligned.fasta', 'output/h1n1pdm/np/nextalign.aligned.fasta', 'output/vic/ha/nextalign.aligned.fasta', 'output/h3n2/na/nextalign.aligned.fasta', 'output/h1n1pdm/ns/nextalign.aligned.fasta', 'output/vic/na/nextalign.aligned.fasta', 'output/h3n2/pa/nextalign.aligned.fasta', 'output/h3n2/ha/nextalign.aligned.fasta', 'output/h1n1pdm/ma/nextalign.aligned.fasta', 'output/h1n1pdm/ha/nextalign.aligned.fasta', 'output/h1n1pdm/pa/nextalign.aligned.fasta', 'output/vic/np/nextalign.aligned.fasta', 'output/h1n1pdm/pb2/nextalign.aligned.fasta', 'output/h3n2/ma/nextalign.aligned.fasta', 'output/h3n2/pb1/nextalign.aligned.fasta', 'output/vic/pb2/nextalign.aligned.fasta', 'output/h3n2/ns/nextalign.aligned.fasta', 'output/vic/ns/nextalign.aligned.fasta', 'output/h1n1pdm/pb1/nextalign.aligned.fasta', 'output/h1n1pdm/na/nextalign.aligned.fasta', 'output/h3n2/np/nextalign.aligned.fasta', 'output/vic/pb1/nextalign.aligned.fasta', 'output/h3n2/pb2/nextalign.aligned.fasta', 'output/vic/ma/nextalign.aligned.fasta'], 'excel_output': ['output/aa_at_positions_for_vic_PA.xlsx', 'output/aa_at_positions_for_h1n1pdm_NP.xlsx', 'output/aa_at_positions_for_vic_HA1.xlsx', 'output/aa_at_positions_for_h3n2_NA.xlsx', 'output/aa_at_positions_for_h1n1pdm_NS2.xlsx', 'output/aa_at_positions_for_vic_NA.xlsx', 'output/aa_at_positions_for_h3n2_PA.xlsx', 'output/aa_at_positions_for_h3n2_HA1.xlsx', 'output/aa_at_positions_for_h1n1pdm_MA.xlsx', 'output/aa_at_positions_for_h1n1pdm_HA1.xlsx', 'output/aa_at_positions_for_h1n1pdm_PA.xlsx', 'output/aa_at_positions_for_vic_NP.xlsx', 'output/aa_at_positions_for_h1n1pdm_PB2.xlsx', 'output/aa_at_positions_for_h3n2_MA.xlsx', 'output/aa_at_positions_for_h3n2_PB1.xlsx', 'output/aa_at_positions_for_vic_PB2.xlsx', 'output/aa_at_positions_for_h3n2_NS2.xlsx', 'output/aa_at_positions_for_vic_NS2.xlsx', 'output/aa_at_positions_for_h1n1pdm_PB1.xlsx', 'output/aa_at_positions_for_h1n1pdm_NA.xlsx', 'output/aa_at_positions_for_h3n2_NP.xlsx', 'output/aa_at_positions_for_vic_PB1.xlsx', 'output/aa_at_positions_for_h3n2_PB2.xlsx', 'output/aa_at_positions_for_vic_MA.xlsx']}\n",
      "['output/vic/pa/nextalign_gene_PA.translation.fasta', 'output/h1n1pdm/np/nextalign_gene_NP.translation.fasta', 'output/vic/ha/nextalign_gene_HA1.translation.fasta', 'output/h3n2/na/nextalign_gene_NA.translation.fasta', 'output/h1n1pdm/ns/nextalign_gene_NS2.translation.fasta', 'output/vic/na/nextalign_gene_NA.translation.fasta', 'output/h3n2/pa/nextalign_gene_PA.translation.fasta', 'output/h3n2/ha/nextalign_gene_HA1.translation.fasta', 'output/h1n1pdm/ma/nextalign_gene_MA.translation.fasta', 'output/h1n1pdm/ha/nextalign_gene_HA1.translation.fasta', 'output/h1n1pdm/pa/nextalign_gene_PA.translation.fasta', 'output/vic/np/nextalign_gene_NP.translation.fasta', 'output/h1n1pdm/pb2/nextalign_gene_PB2.translation.fasta', 'output/h3n2/ma/nextalign_gene_MA.translation.fasta', 'output/h3n2/pb1/nextalign_gene_PB1.translation.fasta', 'output/vic/pb2/nextalign_gene_PB2.translation.fasta', 'output/h3n2/ns/nextalign_gene_NS2.translation.fasta', 'output/vic/ns/nextalign_gene_NS2.translation.fasta', 'output/h1n1pdm/pb1/nextalign_gene_PB1.translation.fasta', 'output/h1n1pdm/na/nextalign_gene_NA.translation.fasta', 'output/h3n2/np/nextalign_gene_NP.translation.fasta', 'output/vic/pb1/nextalign_gene_PB1.translation.fasta', 'output/h3n2/pb2/nextalign_gene_PB2.translation.fasta', 'output/vic/ma/nextalign_gene_MA.translation.fasta']\n",
      "['output/vic/pa/nextalign.aligned.fasta', 'output/h1n1pdm/np/nextalign.aligned.fasta', 'output/vic/ha/nextalign.aligned.fasta', 'output/h3n2/na/nextalign.aligned.fasta', 'output/h1n1pdm/ns/nextalign.aligned.fasta', 'output/vic/na/nextalign.aligned.fasta', 'output/h3n2/pa/nextalign.aligned.fasta', 'output/h3n2/ha/nextalign.aligned.fasta', 'output/h1n1pdm/ma/nextalign.aligned.fasta', 'output/h1n1pdm/ha/nextalign.aligned.fasta', 'output/h1n1pdm/pa/nextalign.aligned.fasta', 'output/vic/np/nextalign.aligned.fasta', 'output/h1n1pdm/pb2/nextalign.aligned.fasta', 'output/h3n2/ma/nextalign.aligned.fasta', 'output/h3n2/pb1/nextalign.aligned.fasta', 'output/vic/pb2/nextalign.aligned.fasta', 'output/h3n2/ns/nextalign.aligned.fasta', 'output/vic/ns/nextalign.aligned.fasta', 'output/h1n1pdm/pb1/nextalign.aligned.fasta', 'output/h1n1pdm/na/nextalign.aligned.fasta', 'output/h3n2/np/nextalign.aligned.fasta', 'output/vic/pb1/nextalign.aligned.fasta', 'output/h3n2/pb2/nextalign.aligned.fasta', 'output/vic/ma/nextalign.aligned.fasta']\n",
      "['output/aa_at_positions_for_vic_PA.xlsx', 'output/aa_at_positions_for_h1n1pdm_NP.xlsx', 'output/aa_at_positions_for_vic_HA1.xlsx', 'output/aa_at_positions_for_h3n2_NA.xlsx', 'output/aa_at_positions_for_h1n1pdm_NS2.xlsx', 'output/aa_at_positions_for_vic_NA.xlsx', 'output/aa_at_positions_for_h3n2_PA.xlsx', 'output/aa_at_positions_for_h3n2_HA1.xlsx', 'output/aa_at_positions_for_h1n1pdm_MA.xlsx', 'output/aa_at_positions_for_h1n1pdm_HA1.xlsx', 'output/aa_at_positions_for_h1n1pdm_PA.xlsx', 'output/aa_at_positions_for_vic_NP.xlsx', 'output/aa_at_positions_for_h1n1pdm_PB2.xlsx', 'output/aa_at_positions_for_h3n2_MA.xlsx', 'output/aa_at_positions_for_h3n2_PB1.xlsx', 'output/aa_at_positions_for_vic_PB2.xlsx', 'output/aa_at_positions_for_h3n2_NS2.xlsx', 'output/aa_at_positions_for_vic_NS2.xlsx', 'output/aa_at_positions_for_h1n1pdm_PB1.xlsx', 'output/aa_at_positions_for_h1n1pdm_NA.xlsx', 'output/aa_at_positions_for_h3n2_NP.xlsx', 'output/aa_at_positions_for_vic_PB1.xlsx', 'output/aa_at_positions_for_h3n2_PB2.xlsx', 'output/aa_at_positions_for_vic_MA.xlsx']\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "data_dir=Path('input')\n",
    "# prefixes=[p.stem for p in data_dir.glob('sequences_*_*.fasta')]\n",
    "\n",
    "# Lineages and Segments in input files for Nextalign or Nextclade.\n",
    "# lineages = ['h1n1pdm', 'h3n2', 'vic', 'yam']\n",
    "# segments = ['pb2', 'pb1', 'pa', 'ha', 'np', 'na', 'ma', 'ns']\n",
    "\n",
    "\n",
    "# Genes in output files from Nextalign or Nextclade.\n",
    "# genes    = ['PB2', 'PB1', 'PA', 'HA1', 'NP', 'NA', 'MA', 'NS2'] # any() ????\n",
    "gene_lookup = {'pb2':'PB2', 'pb1':'PB1', 'pa':'PA', 'ha':'HA1', 'np':'NP', 'na':'NA', 'ma':'MA', 'ns':'NS2'}\n",
    "\n",
    "\n",
    "def input_func_all():\n",
    "    l,l2,l3 = [],[],[]\n",
    "    for path in list(data_dir.glob(\"*.fasta\")):\n",
    "        _, lineage, segment = path.stem.split('_') # sequences_h1n1pdm_ha.fasta\n",
    "        l.append(f\"output/{lineage}/{segment}/nextalign_gene_{gene_lookup[segment]}.translation.fasta\")\n",
    "        l2.append(f\"output/{lineage}/{segment}/nextalign.aligned.fasta\" )\n",
    "        l3.append(f\"output/aa_at_positions_for_{lineage}_{gene_lookup[segment]}.xlsx\")\n",
    "    # Return a dictionary\n",
    "    return {\"genes_translated\" : l, \"nt_segments_aligned\" : l2, \"excel_output\" : l3}\n",
    "\n",
    "rule_all_input_dict = input_func_all()\n",
    "\n",
    "\n",
    "print(rule_all_input_dict)\n",
    "print(rule_all_input_dict[\"genes_translated\"])\n",
    "print(rule_all_input_dict[\"nt_segments_aligned\"])\n",
    "print(rule_all_input_dict[\"excel_output\"])"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Flatten the output to a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['output/vic/pa/nextalign_gene_PA.translation.fasta',\n",
       " 'output/vic/pa/nextalign.aligned.fasta',\n",
       " 'output/aa_at_positions_for_vic_PA.xlsx']"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "data_dir=Path('input')\n",
    "\n",
    "gene_lookup = {'pb2':'PB2', 'pb1':'PB1', 'pa':'PA', 'ha':'HA1', 'np':'NP', 'na':'NA', 'ma':'MA', 'ns':'NS2'}\n",
    "\n",
    "def input_func_all():\n",
    "    l,l2,l3 = [],[],[]\n",
    "    for path in list(data_dir.glob(\"*.fasta\"))[:1]:\n",
    "        _, lineage, segment = path.stem.split('_') # sequences_h1n1pdm_ha.fasta\n",
    "        l.append(f\"output/{lineage}/{segment}/nextalign_gene_{gene_lookup[segment]}.translation.fasta\")\n",
    "        l2.append(f\"output/{lineage}/{segment}/nextalign.aligned.fasta\" )\n",
    "        l3.append(f\"output/aa_at_positions_for_{lineage}_{gene_lookup[segment]}.xlsx\")\n",
    "    return {\"genes_translated\" : l, \"nt_segments_aligned\" : l2, \"excel_output\" : l3}\n",
    "\n",
    "\n",
    "original_list = input_func_all().values()\n",
    "flat_list_of_output_files = [item for l in original_list for item in l]\n",
    "flat_list_of_output_files\n",
    "\n",
    "# input_func_all().values()\n",
    "\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Added gene variable that might be used as wildcard further along..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'genes_translated': ['output/vic/pa/nextalign_gene_PA.translation.fasta', 'output/h1n1pdm/np/nextalign_gene_NP.translation.fasta', 'output/vic/ha/nextalign_gene_HA1.translation.fasta', 'output/h3n2/na/nextalign_gene_NA.translation.fasta', 'output/h1n1pdm/ns/nextalign_gene_NS2.translation.fasta', 'output/vic/na/nextalign_gene_NA.translation.fasta', 'output/h3n2/pa/nextalign_gene_PA.translation.fasta', 'output/h3n2/ha/nextalign_gene_HA1.translation.fasta', 'output/h1n1pdm/ma/nextalign_gene_MA.translation.fasta', 'output/h1n1pdm/ha/nextalign_gene_HA1.translation.fasta', 'output/h1n1pdm/pa/nextalign_gene_PA.translation.fasta', 'output/vic/np/nextalign_gene_NP.translation.fasta', 'output/h1n1pdm/pb2/nextalign_gene_PB2.translation.fasta', 'output/h3n2/ma/nextalign_gene_MA.translation.fasta', 'output/h3n2/pb1/nextalign_gene_PB1.translation.fasta', 'output/vic/pb2/nextalign_gene_PB2.translation.fasta', 'output/h3n2/ns/nextalign_gene_NS2.translation.fasta', 'output/vic/ns/nextalign_gene_NS2.translation.fasta', 'output/h1n1pdm/pb1/nextalign_gene_PB1.translation.fasta', 'output/h1n1pdm/na/nextalign_gene_NA.translation.fasta', 'output/h3n2/np/nextalign_gene_NP.translation.fasta', 'output/vic/pb1/nextalign_gene_PB1.translation.fasta', 'output/h3n2/pb2/nextalign_gene_PB2.translation.fasta', 'output/vic/ma/nextalign_gene_MA.translation.fasta'], 'nt_segments_aligned': ['output/vic/pa/nextalign.aligned.fasta', 'output/h1n1pdm/np/nextalign.aligned.fasta', 'output/vic/ha/nextalign.aligned.fasta', 'output/h3n2/na/nextalign.aligned.fasta', 'output/h1n1pdm/ns/nextalign.aligned.fasta', 'output/vic/na/nextalign.aligned.fasta', 'output/h3n2/pa/nextalign.aligned.fasta', 'output/h3n2/ha/nextalign.aligned.fasta', 'output/h1n1pdm/ma/nextalign.aligned.fasta', 'output/h1n1pdm/ha/nextalign.aligned.fasta', 'output/h1n1pdm/pa/nextalign.aligned.fasta', 'output/vic/np/nextalign.aligned.fasta', 'output/h1n1pdm/pb2/nextalign.aligned.fasta', 'output/h3n2/ma/nextalign.aligned.fasta', 'output/h3n2/pb1/nextalign.aligned.fasta', 'output/vic/pb2/nextalign.aligned.fasta', 'output/h3n2/ns/nextalign.aligned.fasta', 'output/vic/ns/nextalign.aligned.fasta', 'output/h1n1pdm/pb1/nextalign.aligned.fasta', 'output/h1n1pdm/na/nextalign.aligned.fasta', 'output/h3n2/np/nextalign.aligned.fasta', 'output/vic/pb1/nextalign.aligned.fasta', 'output/h3n2/pb2/nextalign.aligned.fasta', 'output/vic/ma/nextalign.aligned.fasta'], 'excel_output': ['output/aa_at_positions_for_vic_PA.xlsx', 'output/aa_at_positions_for_h1n1pdm_NP.xlsx', 'output/aa_at_positions_for_vic_HA1.xlsx', 'output/aa_at_positions_for_h3n2_NA.xlsx', 'output/aa_at_positions_for_h1n1pdm_NS2.xlsx', 'output/aa_at_positions_for_vic_NA.xlsx', 'output/aa_at_positions_for_h3n2_PA.xlsx', 'output/aa_at_positions_for_h3n2_HA1.xlsx', 'output/aa_at_positions_for_h1n1pdm_MA.xlsx', 'output/aa_at_positions_for_h1n1pdm_HA1.xlsx', 'output/aa_at_positions_for_h1n1pdm_PA.xlsx', 'output/aa_at_positions_for_vic_NP.xlsx', 'output/aa_at_positions_for_h1n1pdm_PB2.xlsx', 'output/aa_at_positions_for_h3n2_MA.xlsx', 'output/aa_at_positions_for_h3n2_PB1.xlsx', 'output/aa_at_positions_for_vic_PB2.xlsx', 'output/aa_at_positions_for_h3n2_NS2.xlsx', 'output/aa_at_positions_for_vic_NS2.xlsx', 'output/aa_at_positions_for_h1n1pdm_PB1.xlsx', 'output/aa_at_positions_for_h1n1pdm_NA.xlsx', 'output/aa_at_positions_for_h3n2_NP.xlsx', 'output/aa_at_positions_for_vic_PB1.xlsx', 'output/aa_at_positions_for_h3n2_PB2.xlsx', 'output/aa_at_positions_for_vic_MA.xlsx']}\n",
      "['output/vic/pa/nextalign_gene_PA.translation.fasta', 'output/h1n1pdm/np/nextalign_gene_NP.translation.fasta', 'output/vic/ha/nextalign_gene_HA1.translation.fasta', 'output/h3n2/na/nextalign_gene_NA.translation.fasta', 'output/h1n1pdm/ns/nextalign_gene_NS2.translation.fasta', 'output/vic/na/nextalign_gene_NA.translation.fasta', 'output/h3n2/pa/nextalign_gene_PA.translation.fasta', 'output/h3n2/ha/nextalign_gene_HA1.translation.fasta', 'output/h1n1pdm/ma/nextalign_gene_MA.translation.fasta', 'output/h1n1pdm/ha/nextalign_gene_HA1.translation.fasta', 'output/h1n1pdm/pa/nextalign_gene_PA.translation.fasta', 'output/vic/np/nextalign_gene_NP.translation.fasta', 'output/h1n1pdm/pb2/nextalign_gene_PB2.translation.fasta', 'output/h3n2/ma/nextalign_gene_MA.translation.fasta', 'output/h3n2/pb1/nextalign_gene_PB1.translation.fasta', 'output/vic/pb2/nextalign_gene_PB2.translation.fasta', 'output/h3n2/ns/nextalign_gene_NS2.translation.fasta', 'output/vic/ns/nextalign_gene_NS2.translation.fasta', 'output/h1n1pdm/pb1/nextalign_gene_PB1.translation.fasta', 'output/h1n1pdm/na/nextalign_gene_NA.translation.fasta', 'output/h3n2/np/nextalign_gene_NP.translation.fasta', 'output/vic/pb1/nextalign_gene_PB1.translation.fasta', 'output/h3n2/pb2/nextalign_gene_PB2.translation.fasta', 'output/vic/ma/nextalign_gene_MA.translation.fasta']\n",
      "['output/vic/pa/nextalign.aligned.fasta', 'output/h1n1pdm/np/nextalign.aligned.fasta', 'output/vic/ha/nextalign.aligned.fasta', 'output/h3n2/na/nextalign.aligned.fasta', 'output/h1n1pdm/ns/nextalign.aligned.fasta', 'output/vic/na/nextalign.aligned.fasta', 'output/h3n2/pa/nextalign.aligned.fasta', 'output/h3n2/ha/nextalign.aligned.fasta', 'output/h1n1pdm/ma/nextalign.aligned.fasta', 'output/h1n1pdm/ha/nextalign.aligned.fasta', 'output/h1n1pdm/pa/nextalign.aligned.fasta', 'output/vic/np/nextalign.aligned.fasta', 'output/h1n1pdm/pb2/nextalign.aligned.fasta', 'output/h3n2/ma/nextalign.aligned.fasta', 'output/h3n2/pb1/nextalign.aligned.fasta', 'output/vic/pb2/nextalign.aligned.fasta', 'output/h3n2/ns/nextalign.aligned.fasta', 'output/vic/ns/nextalign.aligned.fasta', 'output/h1n1pdm/pb1/nextalign.aligned.fasta', 'output/h1n1pdm/na/nextalign.aligned.fasta', 'output/h3n2/np/nextalign.aligned.fasta', 'output/vic/pb1/nextalign.aligned.fasta', 'output/h3n2/pb2/nextalign.aligned.fasta', 'output/vic/ma/nextalign.aligned.fasta']\n",
      "['output/aa_at_positions_for_vic_PA.xlsx', 'output/aa_at_positions_for_h1n1pdm_NP.xlsx', 'output/aa_at_positions_for_vic_HA1.xlsx', 'output/aa_at_positions_for_h3n2_NA.xlsx', 'output/aa_at_positions_for_h1n1pdm_NS2.xlsx', 'output/aa_at_positions_for_vic_NA.xlsx', 'output/aa_at_positions_for_h3n2_PA.xlsx', 'output/aa_at_positions_for_h3n2_HA1.xlsx', 'output/aa_at_positions_for_h1n1pdm_MA.xlsx', 'output/aa_at_positions_for_h1n1pdm_HA1.xlsx', 'output/aa_at_positions_for_h1n1pdm_PA.xlsx', 'output/aa_at_positions_for_vic_NP.xlsx', 'output/aa_at_positions_for_h1n1pdm_PB2.xlsx', 'output/aa_at_positions_for_h3n2_MA.xlsx', 'output/aa_at_positions_for_h3n2_PB1.xlsx', 'output/aa_at_positions_for_vic_PB2.xlsx', 'output/aa_at_positions_for_h3n2_NS2.xlsx', 'output/aa_at_positions_for_vic_NS2.xlsx', 'output/aa_at_positions_for_h1n1pdm_PB1.xlsx', 'output/aa_at_positions_for_h1n1pdm_NA.xlsx', 'output/aa_at_positions_for_h3n2_NP.xlsx', 'output/aa_at_positions_for_vic_PB1.xlsx', 'output/aa_at_positions_for_h3n2_PB2.xlsx', 'output/aa_at_positions_for_vic_MA.xlsx']\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "data_dir=Path('input')\n",
    "\n",
    "gene_lookup = {'pb2':'PB2', 'pb1':'PB1', 'pa':'PA', 'ha':'HA1', 'np':'NP', 'na':'NA', 'ma':'MA', 'ns':'NS2'}\n",
    "\n",
    "def input_func_all():\n",
    "    l,l2,l3 = [],[],[]\n",
    "    for path in list(data_dir.glob(\"*.fasta\")):\n",
    "        _, lineage, segment = path.stem.split('_') # sequences_h1n1pdm_ha.fasta\n",
    "        # Get gene variable from dict.\n",
    "        gene = gene_lookup[segment]\n",
    "        l.append(f\"output/{lineage}/{segment}/nextalign_gene_{gene}.translation.fasta\")\n",
    "        l2.append(f\"output/{lineage}/{segment}/nextalign.aligned.fasta\" )\n",
    "        l3.append(f\"output/aa_at_positions_for_{lineage}_{gene}.xlsx\")\n",
    "    # Return a dictionary\n",
    "    return {\"genes_translated\" : l, \"nt_segments_aligned\" : l2, \"excel_output\" : l3}\n",
    "\n",
    "rule_all_input_dict = input_func_all()\n",
    "\n",
    "\n",
    "print(rule_all_input_dict)\n",
    "print(rule_all_input_dict[\"genes_translated\"])\n",
    "print(rule_all_input_dict[\"nt_segments_aligned\"])\n",
    "print(rule_all_input_dict[\"excel_output\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (1445278234.py, line 20)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Input \u001b[0;32mIn [23]\u001b[0;36m\u001b[0m\n\u001b[0;31m    for segment == 'ha' in l5:\u001b[0m\n\u001b[0m                ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "data_dir=Path('input')\n",
    "\n",
    "gene_lookup = {'pb2':'PB2', 'pb1':'PB1', 'pa':'PA', 'ha':'HA1', 'np':'NP', 'na':'NA', 'ma':'M1', 'ns':'NS2'}\n",
    "\n",
    "def input_func_all():\n",
    "    l,l2,l3,l4,l5,l6 = [],[],[],[],[],[]\n",
    "    for path in list(data_dir.glob(\"*.fasta\")):\n",
    "        _, lineage, segment = path.stem.split('_') # sequences_h1n1pdm_ha.fasta\n",
    "        # Get gene variable from dict.\n",
    "        gene = gene_lookup[segment]\n",
    "        l.append(f\"output/{lineage}/{segment}/nextalign_gene_{gene}.translation.fasta\")\n",
    "        l2.append(f\"output/{lineage}/{segment}/nextalign.aligned.fasta\" )\n",
    "        l3.append(f\"output/{lineage}/{segment}/aa_at_positions_for_{gene}.xlsx\")\n",
    "        l4.append(gene)\n",
    "        l5.append(segment)\n",
    "        l6.append(lineage)\n",
    "    # Return a dictionary\n",
    "    return {\"genes_translated\" : l,\n",
    "            \"nt_segments_aligned\" : l2,\n",
    "            \"excel_output\" : l3,\n",
    "            \"gene\": l4,\n",
    "            \"segment\": l5,\n",
    "            \"lineage\": l6}\n",
    "\n",
    "\n",
    "rule_all_input_dict = input_func_all()\n",
    "\n",
    "segments = rule_all_input_dict[\"segment\"]\n",
    "\n",
    "print(segments)\n",
    "\n",
    "lineages = rule_all_input_dict[\"lineage\"]\n",
    "\n",
    "genes = rule_all_input_dict[\"gene\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['pa', 'np', 'ha', 'na', 'ns', 'na', 'pa', 'ha', 'ma', 'ha', 'pa', 'np', 'pb2', 'ma', 'pb1', 'pb2', 'ns', 'ns', 'pb1', 'na', 'np', 'pb1', 'pb2', 'ma']\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "data_dir=Path('input')\n",
    "\n",
    "gene_lookup = {'pb2':'PB2', 'pb1':'PB1', 'pa':'PA', 'ha':'HA1', 'np':'NP', 'na':'NA', 'ma':'M1', 'ns':'NS2'}\n",
    "\n",
    "def input_func_all():\n",
    "    l,l2,l3,l4,l5,l6 = [],[],[],[],[],[]\n",
    "    for path in list(data_dir.glob(\"*.fasta\")):\n",
    "        _, lineage, segment = path.stem.split('_') # sequences_h1n1pdm_ha.fasta\n",
    "        # Get gene variable from dict.\n",
    "        gene = gene_lookup[segment]\n",
    "        l.append(f\"output/{lineage}/{segment}/nextalign_gene_{gene}.translation.fasta\")\n",
    "        l2.append(f\"output/{lineage}/{segment}/nextalign.aligned.fasta\" )\n",
    "        l3.append(f\"output/{lineage}/{segment}/aa_at_positions_for_{gene}.xlsx\")\n",
    "        l4.append(gene)\n",
    "        l5.append(segment)\n",
    "        l6.append(lineage)\n",
    "        # Return a dictionary\n",
    "    return {\"genes_translated\" : l,\n",
    "            \"nt_segments_aligned\" : l2,\n",
    "            \"excel_output\" : l3,\n",
    "            \"gene\": l4,\n",
    "            \"segment\": l5,\n",
    "            \"lineage\": l6}\n",
    "\n",
    "rule_all_input_dict = input_func_all()\n",
    "\n",
    "segments = rule_all_input_dict[\"segment\"]\n",
    "\n",
    "print(segments)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['input/sequences_vic_ha.fasta', 'input/sequences_h3n2_na.fasta', 'input/sequences_vic_na.fasta', 'input/sequences_h3n2_ha.fasta', 'input/sequences_h1n1pdm_ha.fasta', 'input/sequences_h1n1pdm_na.fasta']\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "'PosixPath' object is not iterable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/Users/nic/NIC/Berlin/project_outline.ipynb Cell 15\u001b[0m in \u001b[0;36m<cell line: 28>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/nic/NIC/Berlin/project_outline.ipynb#X20sZmlsZQ%3D%3D?line=24'>25</a>\u001b[0m             filtered_foldernames\u001b[39m.\u001b[39mappend(\u001b[39mstr\u001b[39m(path))  \u001b[39m# Convert the Path object to a string\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/nic/NIC/Berlin/project_outline.ipynb#X20sZmlsZQ%3D%3D?line=25'>26</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m filtered_foldernames\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/nic/NIC/Berlin/project_outline.ipynb#X20sZmlsZQ%3D%3D?line=27'>28</a>\u001b[0m ha_and_na_data_folders \u001b[39m=\u001b[39m data_func_by_segment([\u001b[39m'\u001b[39;49m\u001b[39mha\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39mna\u001b[39;49m\u001b[39m'\u001b[39;49m])\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/nic/NIC/Berlin/project_outline.ipynb#X20sZmlsZQ%3D%3D?line=28'>29</a>\u001b[0m \u001b[39mprint\u001b[39m(ha_and_na_data_folders)\n",
      "\u001b[1;32m/Users/nic/NIC/Berlin/project_outline.ipynb Cell 15\u001b[0m in \u001b[0;36mdata_func_by_segment\u001b[0;34m(segment_list)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/nic/NIC/Berlin/project_outline.ipynb#X20sZmlsZQ%3D%3D?line=19'>20</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mdata_func_by_segment\u001b[39m(segment_list):\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/nic/NIC/Berlin/project_outline.ipynb#X20sZmlsZQ%3D%3D?line=20'>21</a>\u001b[0m     filtered_foldernames \u001b[39m=\u001b[39m []\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/nic/NIC/Berlin/project_outline.ipynb#X20sZmlsZQ%3D%3D?line=21'>22</a>\u001b[0m     \u001b[39mfor\u001b[39;00m path \u001b[39min\u001b[39;00m \u001b[39mlist\u001b[39;49m(data_dir):\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/nic/NIC/Berlin/project_outline.ipynb#X20sZmlsZQ%3D%3D?line=22'>23</a>\u001b[0m         _, lineage, segment \u001b[39m=\u001b[39m path\u001b[39m.\u001b[39mstem\u001b[39m.\u001b[39msplit(\u001b[39m'\u001b[39m\u001b[39m_\u001b[39m\u001b[39m'\u001b[39m)  \u001b[39m# sequences_h1n1pdm_ha.fasta\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/nic/NIC/Berlin/project_outline.ipynb#X20sZmlsZQ%3D%3D?line=23'>24</a>\u001b[0m         \u001b[39mif\u001b[39;00m segment \u001b[39min\u001b[39;00m segment_list:\n",
      "\u001b[0;31mTypeError\u001b[0m: 'PosixPath' object is not iterable"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "input_dir = Path('input')\n",
    "data_dir = Path('data')\n",
    "\n",
    "gene_lookup = {'pb2': 'PB2', 'pb1': 'PB1', 'pa': 'PA', 'ha': 'HA1', 'np': 'NP', 'na': 'NA', 'ma': 'M1', 'ns': 'NS2'}\n",
    "\n",
    "def input_func_by_segment(segment_list):\n",
    "    filtered_filenames = []\n",
    "    for path in list(input_dir.glob(\"*.fasta\")):\n",
    "        _, lineage, segment = path.stem.split('_')  # sequences_h1n1pdm_ha.fasta\n",
    "        if segment in segment_list:\n",
    "            filtered_filenames.append(str(path))  # Convert the Path object to a string\n",
    "    return filtered_filenames\n",
    "\n",
    "ha_and_na_filenames = input_func_by_segment(['ha', 'na'])\n",
    "print(ha_and_na_filenames)\n",
    "\n",
    "\n",
    "def data_func_by_segment(segment_list):\n",
    "    filtered_foldernames = []\n",
    "    for path in list(data_dir):\n",
    "        _, lineage, segment = path.stem.split('_')  # sequences_h1n1pdm_ha.fasta\n",
    "        if segment in segment_list:\n",
    "            filtered_foldernames.append(str(path))  # Convert the Path object to a string\n",
    "    return filtered_foldernames\n",
    "\n",
    "ha_and_na_data_folders = data_func_by_segment(['ha', 'na'])\n",
    "print(ha_and_na_data_folders)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['input/sequences_vic_ha.fasta', 'input/sequences_h3n2_na.fasta', 'input/sequences_vic_na.fasta', 'input/sequences_h3n2_ha.fasta', 'input/sequences_h1n1pdm_ha.fasta', 'input/sequences_h1n1pdm_na.fasta']\n",
      "['_nextclade/data/flu_h3n2_ha', '_nextclade/data/flu_h1n1pdm_na', '_nextclade/data/flu_vic_ha', '_nextclade/data/flu_yam_ha', '_nextclade/data/flu_h1n1pdm_ha', '_nextclade/data/flu_h3n2_na', '_nextclade/data/flu_vic_na']\n"
     ]
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "input_dir = Path('input')\n",
    "\n",
    "gene_lookup = {'pb2': 'PB2', 'pb1': 'PB1', 'pa': 'PA', 'ha': 'HA1', 'np': 'NP', 'na': 'NA', 'ma': 'M1', 'ns': 'NS2'}\n",
    "\n",
    "def input_func_by_segment(segment_list):\n",
    "    filtered_filenames = []\n",
    "    for path in list(input_dir.glob(\"*.fasta\")):\n",
    "        _, lineage, segment = path.stem.split('_')  # sequences_h1n1pdm_ha.fasta\n",
    "        if segment in segment_list:\n",
    "            filtered_filenames.append(str(path))  # Convert the Path object to a string\n",
    "    return filtered_filenames\n",
    "\n",
    "ha_and_na_filenames = input_func_by_segment(['ha', 'na'])\n",
    "print(ha_and_na_filenames)\n",
    "\n",
    "\n",
    "import os\n",
    "data_dir = '_nextclade/data'\n",
    "\n",
    "def data_func_by_segment(segment_list):\n",
    "    filtered_folders = []\n",
    "    for root, dirs, files in os.walk(data_dir):\n",
    "        for folder in dirs:\n",
    "            _, lineage, segment = folder.split('_')\n",
    "            if segment in segment_list:\n",
    "                filtered_folders.append(os.path.join(root, folder))\n",
    "    return filtered_folders\n",
    "\n",
    "ha_and_na_data_folders = data_func_by_segment(['ha', 'na'])\n",
    "print(ha_and_na_data_folders)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'genes_translated': ['output/vic/ha/nextalign_gene_HA1.translation.fasta', 'output/h3n2/na/nextalign_gene_NA.translation.fasta', 'output/vic/na/nextalign_gene_NA.translation.fasta', 'output/h3n2/ha/nextalign_gene_HA1.translation.fasta', 'output/h1n1pdm/ha/nextalign_gene_HA1.translation.fasta', 'output/h1n1pdm/na/nextalign_gene_NA.translation.fasta'], 'nt_segments_aligned': ['output/vic/ha/nextalign.aligned.fasta', 'output/h3n2/na/nextalign.aligned.fasta', 'output/vic/na/nextalign.aligned.fasta', 'output/h3n2/ha/nextalign.aligned.fasta', 'output/h1n1pdm/ha/nextalign.aligned.fasta', 'output/h1n1pdm/na/nextalign.aligned.fasta'], 'excel_output': ['output/vic/ha/aa_at_positions_for_HA1.xlsx', 'output/h3n2/na/aa_at_positions_for_NA.xlsx', 'output/vic/na/aa_at_positions_for_NA.xlsx', 'output/h3n2/ha/aa_at_positions_for_HA1.xlsx', 'output/h1n1pdm/ha/aa_at_positions_for_HA1.xlsx', 'output/h1n1pdm/na/aa_at_positions_for_NA.xlsx'], 'gene': ['HA1', 'NA', 'NA', 'HA1', 'HA1', 'NA'], 'segment': ['ha', 'na', 'na', 'ha', 'ha', 'na'], 'lineage': ['vic', 'h3n2', 'vic', 'h3n2', 'h1n1pdm', 'h1n1pdm']}\n",
      "['HA1', 'NA', 'NA', 'HA1', 'HA1', 'NA']\n",
      "['ha', 'na', 'na', 'ha', 'ha', 'na']\n",
      "['vic', 'h3n2', 'vic', 'h3n2', 'h1n1pdm', 'h1n1pdm']\n"
     ]
    }
   ],
   "source": [
    "def input_func_all(segment_list):\n",
    "    l,l2,l3,l4,l5,l6 = [],[],[],[],[],[]\n",
    "    for path in list(input_dir.glob(\"*.fasta\")):\n",
    "        _, lineage, segment = path.stem.split('_') # sequences_h1n1pdm_ha.fasta\n",
    "        # Get gene variable from dict.\n",
    "        if segment in segment_list:\n",
    "            gene = gene_lookup[segment]\n",
    "            l.append(f\"output/{lineage}/{segment}/nextalign_gene_{gene}.translation.fasta\")\n",
    "            l2.append(f\"output/{lineage}/{segment}/nextalign.aligned.fasta\" )\n",
    "            l3.append(f\"output/{lineage}/{segment}/aa_at_positions_for_{gene}.xlsx\")\n",
    "            l4.append(gene)\n",
    "            l5.append(segment)\n",
    "            l6.append(lineage)\n",
    "    # Return a dictionary\n",
    "    return {\"genes_translated\" : l,\n",
    "            \"nt_segments_aligned\" : l2,\n",
    "            \"excel_output\" : l3,\n",
    "            \"gene\": l4,\n",
    "            \"segment\": l5,\n",
    "            \"lineage\": l6}\n",
    "\n",
    "\n",
    "\n",
    "# Dictionary containing the 6 lists defined above in the function `input_func_all()`\n",
    "input_data = input_func_all(['ha', 'na'])\n",
    "print(input_data)\n",
    "\n",
    "# Get the dictionary values as lists to expand on using zip later on...\n",
    "GENES = input_data[\"gene\"]\n",
    "SEGMENTS = input_data[\"segment\"]\n",
    "LINEAGES = input_data[\"lineage\"]\n",
    "\n",
    "print(GENES)\n",
    "print(SEGMENTS)\n",
    "print(LINEAGES)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['SigPep', 'HA1', 'HA2'], ['NA'], ['NA'], ['SigPep', 'HA1', 'HA2'], ['SigPep', 'HA1', 'HA2'], ['NA']]\n",
      "['ha', 'na', 'na', 'ha', 'ha', 'na']\n",
      "['vic', 'h3n2', 'vic', 'h3n2', 'h1n1pdm', 'h1n1pdm']\n"
     ]
    }
   ],
   "source": [
    "gene_lookup = {\n",
    "    'ha': ['SigPep', 'HA1', 'HA2'],\n",
    "    'na': ['NA']\n",
    "}\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "input_dir = Path('input')\n",
    "\n",
    "def input_func_all(segment_list):\n",
    "    l,l2,l3,l4,l5,l6 = [],[],[],[],[],[]\n",
    "    for path in list(input_dir.glob(\"*.fasta\")):\n",
    "        _, lineage, segment = path.stem.split('_') # sequences_h1n1pdm_ha.fasta\n",
    "        # Get gene variable from dict.\n",
    "        if segment in segment_list:\n",
    "            gene = gene_lookup[segment]\n",
    "            l.append(f\"output/nextclade/{lineage}/{segment}/nextclade.tsv\")\n",
    "            l2.append(f\"output/nextclade/{lineage}/{segment}/nextclade.aligned.fasta\" )\n",
    "            l3.append(f\"output/nextclade/{lineage}/{segment}/nextclade_gene_{gene}.translation.fasta\")\n",
    "            l4.append(gene)\n",
    "            l5.append(segment)\n",
    "            l6.append(lineage)\n",
    "    # Return a dictionary\n",
    "    return {\"nextclade_tsv\" : l,\n",
    "            \"nextclade_aligned\" : l2,\n",
    "            \"nextclade_gene_translated\" : l3,\n",
    "            \"gene\": l4,\n",
    "            \"segment\": l5,\n",
    "            \"lineage\": l6}\n",
    "\n",
    "# Dictionary containing the 6 lists defined above in the function `input_func_all()`\n",
    "input_data = input_func_all(['ha', 'na'])\n",
    "GENES = input_data[\"gene\"]\n",
    "SEGMENTS = input_data[\"segment\"]\n",
    "LINEAGES = input_data[\"lineage\"]\n",
    "print(GENES)\n",
    "print(SEGMENTS)\n",
    "print(LINEAGES)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we will try to count the values across columns in Pandas DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>clade</th>\n",
       "      <th>Pos 81</th>\n",
       "      <th>Pos 133</th>\n",
       "      <th>Pos 135</th>\n",
       "      <th>Pos 137</th>\n",
       "      <th>Pos 140</th>\n",
       "      <th>Pos 144</th>\n",
       "      <th>Pos 145</th>\n",
       "      <th>Pos 156</th>\n",
       "      <th>Pos 192</th>\n",
       "      <th>Pos 222</th>\n",
       "      <th>Pos 223</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EPI_ISL_16600559|A/Netherlands/12245/2022|A_/_...</td>\n",
       "      <td>3C.2a1b.2a.2b</td>\n",
       "      <td>D</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>S</td>\n",
       "      <td>K</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>H</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EPI_ISL_16700806|A/Netherlands/10125/2023|A_/_...</td>\n",
       "      <td>3C.2a1b.2a.2a.3a</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>S</td>\n",
       "      <td>I</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>F</td>\n",
       "      <td>R</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EPI_ISL_16812568|A/Netherlands/10162/2023|A_/_...</td>\n",
       "      <td>3C.2a1b.2a.2b</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>S</td>\n",
       "      <td>K</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>H</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EPI_ISL_16699238|A/Netherlands/00245/2023|A_/_...</td>\n",
       "      <td>3C.2a1b.2a.2b</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>S</td>\n",
       "      <td>K</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>H</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EPI_ISL_16254955|A/Netherlands/01438/2022|A_/_...</td>\n",
       "      <td>3C.2a1b.2a.2b</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>A</td>\n",
       "      <td>S</td>\n",
       "      <td>K</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>H</td>\n",
       "      <td>I</td>\n",
       "      <td>R</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            Sequence             clade Pos 81   \n",
       "0  EPI_ISL_16600559|A/Netherlands/12245/2022|A_/_...     3C.2a1b.2a.2b      D  \\\n",
       "1  EPI_ISL_16700806|A/Netherlands/10125/2023|A_/_...  3C.2a1b.2a.2a.3a      N   \n",
       "2  EPI_ISL_16812568|A/Netherlands/10162/2023|A_/_...     3C.2a1b.2a.2b      N   \n",
       "3  EPI_ISL_16699238|A/Netherlands/00245/2023|A_/_...     3C.2a1b.2a.2b      N   \n",
       "4  EPI_ISL_16254955|A/Netherlands/01438/2022|A_/_...     3C.2a1b.2a.2b      N   \n",
       "\n",
       "  Pos 133 Pos 135 Pos 137 Pos 140 Pos 144 Pos 145 Pos 156 Pos 192 Pos 222   \n",
       "0       N       A       S       K       S       S       H       I       R  \\\n",
       "1       N       A       S       I       S       S       S       F       R   \n",
       "2       N       A       S       K       S       S       H       I       R   \n",
       "3       N       A       S       K       S       S       H       I       R   \n",
       "4       N       A       S       K       S       S       H       I       R   \n",
       "\n",
       "  Pos 223  \n",
       "0       I  \n",
       "1       V  \n",
       "2       I  \n",
       "3       I  \n",
       "4       I  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_table('temp/positions_clades_aa_h3n2_HA1.txt')\n",
    "df.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Pos 81', 'Pos 133', 'Pos 135', 'Pos 137', 'Pos 140', 'Pos 144',\n",
       "       'Pos 145', 'Pos 156', 'Pos 192', 'Pos 222', 'Pos 223'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos_columns = df.columns[2:]\n",
    "pos_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = df[ pos_columns ].value_counts()\n",
    "counts.to_csv(('temp/positions_clades_aa_h3n2_HA1_counts.csv'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqName</th>\n",
       "      <th>clade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EPI_ISL_17672205|A/Netherlands/01598/2023|A_/_...</td>\n",
       "      <td>6B.1A.5a.2a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EPI_ISL_17672203|A/Netherlands/01584/2023|A_/_...</td>\n",
       "      <td>6B.1A.5a.2a.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EPI_ISL_17672201|A/Netherlands/01578/2023|A_/_...</td>\n",
       "      <td>6B.1A.5a.2a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EPI_ISL_17672199|A/Netherlands/01568/2023|A_/_...</td>\n",
       "      <td>6B.1A.5a.2a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EPI_ISL_17672198|A/Netherlands/01560/2023|A_/_...</td>\n",
       "      <td>6B.1A.5a.2a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>EPI_ISL_15544289|A/Netherlands/11759/2022|A_/_...</td>\n",
       "      <td>6B.1A.5a.2a.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>601</th>\n",
       "      <td>EPI_ISL_14918898|A/Netherlands/11721/2022|A_/_...</td>\n",
       "      <td>6B.1A.5a.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>602</th>\n",
       "      <td>EPI_ISL_15774528|A/Netherlands/11801/2022|A_/_...</td>\n",
       "      <td>6B.1A.5a.2a.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>603</th>\n",
       "      <td>EPI_ISL_15544291|A/Netherlands/11767/2022|A_/_...</td>\n",
       "      <td>6B.1A.5a.2a.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604</th>\n",
       "      <td>EPI_ISL_13324721|A/Netherlands/11651/2022|A_/_...</td>\n",
       "      <td>6B.1A.5a.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>605 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               seqName           clade\n",
       "0    EPI_ISL_17672205|A/Netherlands/01598/2023|A_/_...  6B.1A.5a.2a   \n",
       "1    EPI_ISL_17672203|A/Netherlands/01584/2023|A_/_...  6B.1A.5a.2a.1 \n",
       "2    EPI_ISL_17672201|A/Netherlands/01578/2023|A_/_...  6B.1A.5a.2a   \n",
       "3    EPI_ISL_17672199|A/Netherlands/01568/2023|A_/_...  6B.1A.5a.2a   \n",
       "4    EPI_ISL_17672198|A/Netherlands/01560/2023|A_/_...  6B.1A.5a.2a   \n",
       "..                                                 ...             ...\n",
       "600  EPI_ISL_15544289|A/Netherlands/11759/2022|A_/_...  6B.1A.5a.2a.1 \n",
       "601  EPI_ISL_14918898|A/Netherlands/11721/2022|A_/_...  6B.1A.5a.1    \n",
       "602  EPI_ISL_15774528|A/Netherlands/11801/2022|A_/_...  6B.1A.5a.2a.1 \n",
       "603  EPI_ISL_15544291|A/Netherlands/11767/2022|A_/_...  6B.1A.5a.2a.1 \n",
       "604  EPI_ISL_13324721|A/Netherlands/11651/2022|A_/_...  6B.1A.5a.1    \n",
       "\n",
       "[605 rows x 2 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Import generated Excel file from rule 'get_positions'\n",
    "excel_output = \"output/positions/h1n1pdm/ha/positions_aa_h1n1pdm_HA1.xlsx\"\n",
    "\n",
    "# Import generated csv file from 'nextclade'\n",
    "nextclade_csv = \"output/nextclade/h1n1pdm/ha/nextclade.csv\"\n",
    "\n",
    "# Set the path and name of the final exported Excel output file.\n",
    "excel_with_clades = \"output/positions/h1n1pdm/ha/positions_clades_aa_h1n1pdm_HA1.xlsx\"\n",
    "\n",
    "# Read Excel sheet with pandas and store Dataframe in 'xlsx'\n",
    "xlsx = pd.read_excel(excel_output, sheet_name=\"Sheet1\")\n",
    "\n",
    "# Read csv sheet with pandas and store Dataframe in 'csv'\n",
    "csv = pd.read_csv(nextclade_csv, sep=\";\")\n",
    "csv.columns = csv.columns.str.strip()\n",
    "csv[['seqName', 'clade']]\n",
    "\n",
    "# Get 'seqName' and 'clade' columns from the csv data"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "vcs",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
